### DISCUSSION TITLE
Permit using Click types as annotations

### DISCUSSION THREAD

**Discussion by deepyaman** [fastapi/typer#591] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import click
import typer
import typer.main
from typer.testing import CliRunner


def click_type_pass_through(f):
    @wraps(f)
    def wrapper(*args, **kwds):
        if isinstance(kwds["annotation"], click.ParamType):
            return kwds["annotation"]
        else:
            return f(*args, **kwds)

    return wrapper


typer.main.get_click_type = click_type_pass_through(typer.main.get_click_type)

app = typer.Typer()

def config_file_callback(ctx, param, value):
    ctx.default_map = {"name": {"first": "Camila", "last": "Cabello"}}
    return value

def name_callback(ctx, param, value):
    if isinstance(value, dict):
        return value
    first, last = value.split()
    return {"first": first, "last": last}

@app.command()
def main(
    config: str = typer.Option("--config", callback=config_file_callback),
    name: click.UNPROCESSED = typer.Option("--name", callback=name_callback),
):
    print(f"User: {name['first']} {name['last']}")

runner = CliRunner()
result = runner.invoke(app, args)
assert "User: Camila Cabello" in result.stdout
```

### Description

I want to use a combination of callbacks to define more complex config in a file. For a real-world example using Typer, see https://github.com/deepyaman/kedro-argo/commit/d2eaf1b8f1898293be15b944c88c9eff85e7d6ba. For an example using Click (that I referenced for my Typer implementation), see https://github.com/kedro-org/kedro/blob/5906440/kedro/framework/cli/project.py#L329-L342.

### Wanted Solution

I would like to avoid patching `typer.main.get_click_type` to use the broader set of Click types, including `click.UNPROCESSED`.

### Wanted Code

```python
import click
import typer
from typer.testing import CliRunner

app = typer.Typer()

def config_file_callback(ctx, param, value):
    ctx.default_map = {"name": {"first": "Camila", "last": "Cabello"}}
    return value

def name_callback(ctx, param, value):
    if isinstance(value, dict):
        return value
    first, last = value.split()
    return {"first": first, "last": last}

@app.command()
def main(
    config: str = typer.Option("--config", callback=config_file_callback),
    name: click.UNPROCESSED = typer.Option("--name", callback=name_callback),
):
    print(f"User: {name['first']} {name['last']}")

runner = CliRunner()
result = runner.invoke(app, args)
assert "User: Camila Cabello" in result.stdout
```

### Alternatives

This could be addressed by the broader #443. #446 presents a simpler, more targeted solution.

### Operating System

macOS

### Operating System Details

Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:37 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T6000 arm64

### Typer Version

0.6.1

### Python Version

3.10.5

**Comment by tiangolo** [comment, marked as answer]
Thanks! Support for this was added in https://github.com/tiangolo/typer/pull/583, the new docs are here: https://typer.tiangolo.com/tutorial/parameter-types/custom-types/, it's available in Typer `0.8.0`, just released. :tada:
